src/object.d

```diff
@@ -3241,6 +3241,268 @@ template RTInfo(T)
     enum RTInfo = null;
 }
 
+int __cmp(T1, T2)(T1[] s1, T2[] s2)
+{
+    import core.internal.traits : Unqual;
+    alias U1 = Unqual!T1;
+    alias U2 = Unqual!T2;
+
+    static @trusted ref R at(R)(R[] r, size_t i) { return r.ptr[i]; }
+
+    // objects
+    // Old code went to object.d/TypeInfo_array.compare and then to
+    // object.d/TypeInfo_Class.compare
+    // which combined result in the same code as below.
+    static if (is(U1 : Object) && is(U2 : Object))
+    {
+        auto len = s1.length;
+
+        if (s2.length < len)
+            len = s2.length;
+
+        foreach (const u; 0 .. len)
+        {
+            auto o1 = at(s1, u);
+            auto o2 = at(s2, u);
+
+            if (o1 is o2)
+                continue;
+
+            // Regard null references as always being "less than"
+            if (o1)
+            {
+                if (!o2)
+                    return 1;
+                auto c = o1.opCmp(o2);
+                if (c != 0)
+                    return c < 0 ? -1 : 1;
+            }
+            else
+            {
+                return -1;
+            }
+        }
+        return s1.length < s2.length ? -1 : (s1.length > s2.length);
+    }
+    // floating point types
+    else static if (__traits(isFloating, U1))
+    {
+        static if (is(U1 == ifloat)) alias F = float;
+        else static if (is(U1 == idouble)) alias F = double;
+        else static if (is(U1 == ireal)) alias F = real;
+        else alias F = U1;
+
+        static int compare(F f1, F f2)
```

Comment:
Cool! This is getting ready, but I now realized I gave you the wrong advice when I told you to move these functions inside `__cmp`. This has to do with a subtlety in the way nested static functions are generated: even though you constrained `compare` properly to only accept certain types, the compiler will still generate one instance for each distinct outer template.

I'll approve this (if there are no other issues) and will follow with a fixup.